### PR TITLE
Benchmarks and results

### PR DESCRIPTION
**Pull request by mcarmonaa**

### CODE REVIEW

**Comment by mcuadros** [comment]
@mcarmonaa update the library in different PRs if not is very hard to read the PR

File: benchmark/run-slow-benchmark.sh [hunk 1]
```diff
@@ -0,0 +1,4 @@
+#!/bin/sh
+
+go test -run=NONE -bench=. -slow -benchtime=100ms -timeout=100h >benchmark/output/enry_samples.bench && \
```
Comment (mcuadros) [inline benchmark/run-slow-benchmark.sh]: I prefer move this sh to the makefile

File: benchmark/plot-histogram.sh [hunk 2]
```diff
@@ -0,0 +1,4 @@
+#!/bin/sh
```
Comment (smola) [inline benchmark/plot-histogram.sh]: I think you don't need this file if you use the sheband `#!/usr/bin/env gnuplot` on the gnuplot script.

File: benchmark/plot-histogram.gp [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+set key under
+
+set style data histogram
+set style histogram clustered gap 1 title offset 1,1
+set style fill solid noborder
+set boxwidth 0.95
+set grid y
+set bmargin 12
+set autoscale
+set title "Number of files whose processed time was inside time interval"
```
Comment (smola) [inline benchmark/plot-histogram.gp]: Number of sample files per processing time

File: README.md [hunk 4]
```diff
@@ -33,16 +33,46 @@ fmt.Println(lang)
...
 
     make clean code-generate
 
 To run the tests
 
     make test
 
+Benchmark
+---------
+Enry's language detection has been compared  with Linguist's language detection. In order to do that, linguist's project directory [*linguist/samples*](https://github.com/github/linguist/tree/master/samples) was used as a set of files to run benchmarks against.
```
Comment (abeaumont) [inline README.md]: Minor: Now there's a mix of titles with a blank line afterwards and titles without one. Make it consistent.

File: README.md [hunk 5]
```diff
@@ -33,16 +33,46 @@ fmt.Println(lang)
...
-*enry* re-uses parts of original [linguist](https://github.com/github/linguist) especially data in `languages.yml` to generate internal data structures. In oreder to update to latest upstream run
+*enry* re-uses parts of original [linguist](https://github.com/github/linguist) especially data in `languages.yml` to generate internal data structures. In order to update to latest upstream run
 
     make clean code-generate
 
 To run the tests
 
     make test
 
+Benchmark
```
Comment (abeaumont) [inline README.md]: Minor: `Benchmarks` or `Benchmarking` may be better, since it's not a single benchmark.

File: README.md (hunk 4 above)
Comment (abeaumont) [inline README.md]: Minor: double spaces

File: README.md [hunk 6]
```diff
@@ -33,16 +33,46 @@ fmt.Println(lang)
...
+
+ Following results were obtained:
+
+![histogram](https://raw.githubusercontent.com/src-c/enry/master/benchmark/histogram/distribution.jpg)
+
+The histogram represents the number of files for which spent time in language detection was in the range of the time interval indicated in x axis.
+
+So reviewing the comparison enry/linguist, you can see the most of the files were detected in less time than linguist does.
+
+We detected some few cases enry turns slower than linguist due to Golang regexp while Ruby uses [oniguruma](https://github.com/kkos/oniguruma) as a regular expression engine (a regular expression library written in C).
```
Comment (abeaumont) [inline README.md]: This sentence is a bit awkward. Maybe something along:

```
We detected some few cases enry turns slower than linguist.
This is due to Golang's regexp engine being slower than Ruby's, which uses [oniguruma](https://github.com/kkos/oniguruma) library, written in C.
```

File: README.md [hunk 7]
```diff
@@ -33,16 +33,46 @@ fmt.Println(lang)
...
+
+You can find scripts and additional information (as software and hardware used, and benchmark's results per sample file) in [*benchmark*](https://github.com/src-d/enry/tree/master/benchmark) directory.
+
+If you want to reproduce the same experiment you can run:
+
+    benchmark/run.sh
+
+from the root's project directory and It runs benchmarks for enry and linguist, parse the output, create csv files and create a histogram (you must have installed [gnuplot](http://gnuplot.info) in your system to get the histogram). It can take to much time, so to run local benchmarks to take a quick look you can run either:
+
+    make benchmarks
```
Comment (abeaumont) [inline README.md]: `benchmark` is used in singular in some places and in plural in others. It should be consistent (plural makes more sense I think).

File: README.md [hunk 8]
```diff
@@ -33,16 +33,46 @@ fmt.Println(lang)
...
+
+to get time averages for main detection function and strategies for the whole samples set  or:
```
Comment (abeaumont) [inline README.md]: minor: double spaces

File: .gitignore [hunk 9]
```diff
@@ -1 +1,2 @@
 .linguist
+benchmark/output
```
Comment (abeaumont) [inline .gitignore]: This should be `benchmarks/output` now

File: README.md [hunk 10]
```diff
@@ -1,6 +1,6 @@
 # enry [![GoDoc](https://godoc.org/gopkg.in/src-d/enry.v1?status.svg)](https://godoc.org/gopkg.in/src-d/enry.v1) [![Build Status](https://travis-ci.org/src-d/enry.svg?branch=master)](https://travis-ci.org/src-d/enry) [![codecov](https://codecov.io/gh/src-d/enry/branch/master/graph/badge.svg)](https://codecov.io/gh/src-d/enry)
 
-File programming language detector and toolbox to ignore binary or vendored files. *enry*, started as a port to _Go_ of the original [linguist](https://github.com/github/linguist) _Ruby_ library, that has an improved *performance of 100x*.
+File programming language detector and toolbox to ignore binary or vendored files. *enry*, started as a port to _Go_ of the original [linguist](https://github.com/github/linguist) _Ruby_ library, that has an improved *performance of 200%*.
```
Comment (abeaumont) [inline README.md]: s/performance of 200%/2x performance/?

**Comment by abeaumont** [comment]
@mcarmonaa please rebase once PR is approved

**Comment by abeaumont** [comment]
@mcarmonaa there're some conflicts, please rebase again :S

**Pull request merged**